lib/internal/modules/esm/load.js

```diff
@@ -30,36 +30,36 @@ const {
 const DATA_URL_PATTERN = /^[^/]+\/[^,;]+(?:[^,]*?)(;base64)?,([\s\S]*)$/;
 
 async function getSource(url, context) {
-  const parsed = new URL(url);
-  let responseURL = url;
+  const protocol = url.protocol;
+  let responseURL = url.href;
   let source;
-  if (parsed.protocol === 'file:') {
+  if (protocol === 'file:') {
     const { readFile: readFileAsync } = require('internal/fs/promises').exports;
-    source = await readFileAsync(parsed);
-  } else if (parsed.protocol === 'data:') {
-    const match = RegExpPrototypeExec(DATA_URL_PATTERN, parsed.pathname);
+    source = await readFileAsync(url);
+  } else if (protocol === 'data:') {
+    const match = RegExpPrototypeExec(DATA_URL_PATTERN, url.pathname);
     if (!match) {
-      throw new ERR_INVALID_URL(url);
+      throw new ERR_INVALID_URL(responseURL);
     }
     const { 1: base64, 2: body } = match;
     source = BufferFrom(decodeURIComponent(body), base64 ? 'base64' : 'utf8');
   } else if (experimentalNetworkImports && (
-    parsed.protocol === 'https:' ||
-    parsed.protocol === 'http:'
+    protocol === 'https:' ||
+    protocol === 'http:'
   )) {
     const { fetchModule } = require('internal/modules/esm/fetch_module');
-    const res = await fetchModule(parsed, context);
+    const res = await fetchModule(url, context);
     source = await res.body;
     responseURL = res.resolvedHREF;
   } else {
     const supportedSchemes = ['file', 'data'];
     if (experimentalNetworkImports) {
       ArrayPrototypePush(supportedSchemes, 'http', 'https');
     }
-    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, supportedSchemes);
+    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(url, supportedSchemes);
   }
   if (policy?.manifest) {
-    policy.manifest.assertIntegrity(parsed, source);
+    policy.manifest.assertIntegrity(url, source);
```

Comment:
The first thing `assertIntegrity` does is to convert its first argument to a string https://github.com/nodejs/node/blob/9658d84dddd3835a61e3f270928f3480abadb5ca/lib/internal/policy/manifest.js#L570-L571
We could potentially save some more CPU cycles by keeping it.
```suggestion
    policy.manifest.assertIntegrity(href, source);
```

Reply:
You're right. Updating it. Thanks!